### ISSUE TITLE
Cypress tests slow down or fail, UI is unresponsive in 3.3.1

### ISSUE THREAD

**Issue by joydeep100** [cypress-io/cypress#4411]
### Current behavior:
I noticed the some of our API, mainly XHR requests were continuously getting looped indefinitely when i upgraded Cypress from 3.2.0 to 3.3.0. So we rolled it back.
Now we hit some other bugs, for which we wanted to upgrade to Cypress 3.3.1, once again i am getting a similar issue.

1. For some of the tests some XHR requests are loading indefinitely.
2. The Timer is not incrementing
3. The Test Runner is not responding
 
<img width="501" alt="cypress_1" src="https://user-images.githubusercontent.com/45758076/59108876-936d0180-8959-11e9-9adb-afab8a7e2be1.png">
<img width="1439" alt="cypress_2" src="https://user-images.githubusercontent.com/45758076/59108878-94059800-8959-11e9-9ea7-60a260d8c486.png">
<img width="938" alt="cypress_3" src="https://user-images.githubusercontent.com/45758076/59108879-94059800-8959-11e9-9f1a-63deb1b1a91c.png">

```
------------ Configuration -------
{
animationDistanceThreshold: 5,
baseUrl: null,
blacklistHosts: null,
chromeWebSecurity: true,
defaultCommandTimeout: 20000,
env: {
CLIENT: 'demoaccount',
failOnSnapshotDiff: false},
execTimeout: 60000,
fileServerFolder: '',
fixturesFolder: 'cypress/fixtures',
hosts: null,
integrationFolder: 'cypress/integration',
modifyObstructiveCode: true,
numTestsKeptInMemory: 1,
pageLoadTimeout: 60000,
pluginsFile: 'cypress/plugins',
port: null,
reporter: 'spec',
reporterOptions: null,
requestTimeout: 5000,
responseTimeout: 30000,
screenshotsFolder: 'cypress/screenshots',
supportFile: 'cypress/support',
taskTimeout: 60000,
testFiles: '**/*.{spec.js,feature}',
trashAssetsBeforeRuns: true,
userAgent: null,
video: true,
videoCompression: 32,
videoUploadOnPasses: true,
videosFolder: 'cypress/videos',
viewportHeight: 660,
viewportWidth: 1000,
waitForAnimations: true,
watchForFileChanges: true}
-----------
```

Scenario failing- `Validating the time/date picker yields a correct date in the calendar while scheduling email message (we are using cucumber library)`

### Desired behavior:

Things are fine in 3.2.0, except some known issues like forwarding of socket failures related error in HTTP api level, which causes some parse error (Internal Error 500) in cypress.

### Steps to reproduce: (app code and test code)

I am sharing some snippet of the login code, which we use to login. 

```js
Cypress.Commands.add(
	'login',
	{
		prevSubject: 'true'
	},
	(client, user) => {
		cy.logout();
		return cy.visit(`${BASE_URL}`).then(() => {
			return cy.getCookie('csrftoken').then(({ value }) => {
				return cy
					.request({
						url: `${client.url}/accounts/login`,
						form: true,
						method: 'POST',
						body: {
							username: client.users[user].email,
							password: client.users[user].password,
							csrfmiddlewaretoken: value
						}
					})
					.then(() => client);
			});
		});
	}
);
```

### Versions
cypress - 3.3.1
electron - 61
OS - Mac OS Mojave 10.14.5

### Attachments
[cy_3.3.0_debug.log](https://github.com/cypress-io/cypress/files/3276250/cy_3.3.0_debug.log)
[cy_3.3.1_debug.log](https://github.com/cypress-io/cypress/files/3276251/cy_3.3.1_debug.log)

**Comment by joydeep100** [comment]
One more interesting observation is that, this works fine in Chrome75. This seem to be failing only in Electron browser.

**Comment by jennifer-shehane** [comment]
@joydeep100 If you're using ServiceWorkers, there's a known issue for this: https://github.com/cypress-io/cypress/issues/1716#issuecomment-454161526

This issue was opened concerning WebSockets in 3.3.1 also. https://github.com/cypress-io/cypress/issues/4366

Can you narrow down if this is happening in 3.3.0 / 3.3.1 / or both?

And if you can, run Cypress in [debug mode](https://on.cypress.io/debugging#Debug-the-Command-Line) mode and print the logs here/

**Comment by joydeep100** [comment]
Hi @jennifer-shehane , yes the issue is reproducible in cypress-3.3.0 as well. I have attached the debug logs from both the versions.

**Comment by abzainuddin** [comment]
I'm seeing the same problem. In 3.2.0 everything works fine, but upgrading to 3.3.0 / 3.3.1 some requests seems to get "stuck".

When checking the backend logs, the request is successfully responded, but Cypress seems to miss the response.

Still trying to pinpoint what exactly causes this problem. It seems to only occur with one specific test.

**Comment by flotwig** [comment]
@abzainuddin Can you share the spec code that reproduces this issue?

**Comment by abzainuddin** [comment]
This test most of the time doesn't work, once in a while it does work:

```
it('should be able add success rate to a quotation', () => {
    cy.loadFixture(fixture);
    cy.login('account_manager@codeyellow.nl');
    cy.visit('/orders/project/1/order/1/edit/quotations');

    cy.get('.handshake').click();
    cy.contains('Add').click();
    cy.shouldShowValidationNotification();
    cy.get('[name="rate"]').hasValidationError('Rate must be an integer');

    cy.get('[name="rate"]').type('123');
    cy.contains('Add').click();
    cy.shouldShowValidationNotification();
    cy.get('[name="rate"]').hasValidationError('Rate must be between 0 and 100');

    cy.get('[name="rate"]').type('{selectall}15');
    cy.contains('Add').click(); // Causes infinite pending request, while backend reports done.
    cy.shouldShowSaveNotification();

    cy.contains('15%').should('be.visible');
});
```

After the test has failed (because of timeout), you do see in the dev console that the request is done:

![image](https://user-images.githubusercontent.com/3865520/59741460-59f69900-926b-11e9-9254-df2e6f14e237.png)

See also this screencast https://people.codeyellow.nl/burhan/cypress-bug.mp4 (somehow without my mouse, but you get the picture). After a minute or so the request does resolve. I've tested using different browsers, all giving the same result:

![image](https://user-images.githubusercontent.com/3865520/59741907-9d9dd280-926c-11e9-84b2-dc492a5a2f3b.png)

**Comment by flotwig** [comment]
Okay, I was able to sorta reproduce this by writing tests that do a toooon of mocked route request all at once. Eventually Cypress locks up and starts stuttering and shows the symptoms you're experiencing:

Spec code:

```js
const letters = ['a','b','c','d','e','f']

it('should get responses for all XHRs', function() {
  cy.server()
  cy.pause() // just so i can record a video of it before it locks up
  
  letters.forEach(l => {
    cy.route(`**/${l}`, 'it wokred').as(l + 'g')
    cy.route('POST', `**/${l}`, 'it wokred').as(l + 'p')
  })

  cy.visit('/index-xhrs.html')

  letters.forEach(l => {
    cy.wait(`@${l}g`)
    cy.wait(`@${l}p`)
  })
})
```

`/index-xhrs.html`:

```html
<script type="text/javascript">
const letters = ['a','b','c','d','e','f']

Cypress._.times(200, (i) => {
  setTimeout(() => {
    const xhr = new XMLHttpRequest
    xhr.open('GET', `http://example.com/${letters[i % letters.length]}`)
    xhr.send()

    const xhr2 = new XMLHttpRequest
    xhr2.open('POST', `http://example.com/${letters[i % letters.length]}`)
    xhr2.send()
  }, Math.random() * 5000)
})
</script>
```

Running in Electron, you can clearly see Cypress freezing and struggling to keep up: https://drive.google.com/open?id=1wbC5cobVkvJcA0wL0nVlweaEVT5wRzhN

The same thing happens in Chrome. I actually got it to fail the test because an XHR timed out as well. You can also see that Cypress is using 150% CPU as it tries to keep the tests running: https://drive.google.com/open?id=1G5kugjCn5A2dq04kKyRHpxudU6fjTHJF

To me, this looks like there's a performance issue here. I'm going to test with some of the performance improvements that will be in 3.3.2 and report back if that fixes it.

**Comment by flotwig** [comment]
This slowdown appears to happen with all Cypress commands, @fr0 in #4313 created an example repo (https://github.com/fr0/cypress-slowdown) that shows a 2x slowdown with .get, .click and other commands. I profiled the difference between 3.2.0 and `develop` and it also comes down re-rendering. It looks like everything under every `Command` is re-rendering over and over:

Sample call tree of one update of `Tree` from 3.2.0:
![image](https://user-images.githubusercontent.com/1151760/59951157-0c825380-9446-11e9-8c53-509873570ff2.png)

For comparison, sample call tree of the same update from latest `develop`: 
![image](https://user-images.githubusercontent.com/1151760/59951016-a4337200-9445-11e9-852b-b477b10bf80a.png)

`Command.shouldComponentUpdate` looks like the root of the slowdown, it's no longer returning false when it should and so everything re-renders on every change to `Test`'s state or props.

**Comment by fr0** [comment]
Since it was introduced in 3.3.0, it might help to do a `git bisect` to narrow down the exact commit that introduced the issue.

**Comment by fr0** [comment]
Not sure if this is useful information or not, but I just tried my test case with `--browser chrome` and the slowdown occurred, which seems to contradict the earlier comment that "this works fine in Chrome75. This seem to be failing only in Electron browser."

**Comment by flotwig** [comment]
Yeah, it happens in both Chromium 75 and Electron for me too.

**Comment by flotwig** [comment]
Thanks for reminding me that `git bisect` exists. ;) Tracked it down to #3188, which isn't a lot of changes, picking through it now.

**Comment by fr0** [comment]
Nice! Thanks for your hard work. I verified that the latest build does, in fact, fix my issue.

**Comment by joydeep100** [comment]
Hello @flotwig @jennifer-shehane 

I just wanted to verify that the issue is closed. Hence during my re-testing with version 3.3.2 i can see the same issue is occurring again and i confirm that this occurs only in electron.

One of the calls that (XHR) that kept looping was `mark_as/` . I am attaching a HAR dump file for the network requests during the test. I hope that would be useful.

[cypress.har.txt](https://github.com/cypress-io/cypress/files/3344077/cypress.har.txt)

p.s. Please rename `cypress.har.txt` to `cypress.har` if your HAR analyser needs that.

**Comment by jennifer-shehane** [comment]
@joydeep100 Yes, this issue was intended to be completely fixed in version 3.3.2. 

@flotwig can you take a look at the HAR file?

**Comment by joydeep100** [comment]
Hi @flotwig , Just curious to know if you were able to corner the issue.

**Comment by kris4345** [comment]
I got this thank you

**Comment by flotwig** [comment]
@joydeep100 Let's move discussion of this issue to #4668, since the history of this issue has become about something else. I thought we fixed the issue you're experiencing in 3.3.2, but apparently I reproducing another potential cause of XHR flakiness all along :stuck_out_tongue:
